akka-stream/src/main/scala/akka/stream/impl/io/TcpConnectionStream.scala

```diff
@@ -154,17 +142,25 @@ private[akka] abstract class TcpStreamActor(val settings: ActorMaterializerSetti
     override def complete(): Unit = {
       if (!closed && initialized) {
         closed = true
-        if (tcpInputs.isClosed && (halfClose || lastWriteAcked)) {
-          // We can immediately close if
-          // - half close mode, and read size already finished
-          // - full close mode, and last write has been acked
-          //
-          // if in full close mode, and has a non-acknowledged write, we will do the closing in handleWrite
-          // when the Ack arrives
-          connection ! Close
-          tryShutdown()
-        } else
-          connection ! ConfirmedClose
+
+        if (halfClose) {
+          if (tcpInputs.isClosed) {
+            // Reading has stopped, either because of cancel, or PeerClosed, just Close now
+            connection ! Close
+            tryShutdown()
```

Comment:
These two lines now show up several times. Maybe it should be given a name now... :)